File: Manuel/개인 활동/Class 3/바이러스/BFS/virus_bfs.cpp
#include <iostream>
#include <vector>
#include <queue> // 큐를 사용하기 위한 헤더 파일

using namespace std;

int main()
{
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    int num_computers;
    cin >> num_computers;
    int num_connections;
    cin >> num_connections;
    
    // 그래프를 인접 리스트로 표현.
    vector<vector<int>> graph(num_computers + 1); // 컴퓨터 번호를 1부터 N까지 부여하기 때문에 num_computers + 1 크기로 벡터를 생성

    // 연결 정보 추가
    for(int i = 0; i < num_connections; i++)
    {
        int c1, c2;
        cin >> c1 >> c2;
        // 양방향 연결이므로 양쪽에 정보 추가
        graph[c1].push_back(c2);
        graph[c2].push_back(c1);
    }

    // 방문 여부를 기록할 배열 생성
    // 초기값은 false
    vector<bool> visited(num_computers + 1, false);

    queue<int> q; // BFS를 위한 큐 생성
    q.push(1); // 시작점인 1번을 큐에 넣음
    visited[1] = true; // 방문 표기

    int count_infected = 0; // 바이러스에 걸린 컴퓨터 수

    while(!q.empty())
    {
        int current_computer = q.front(); // 큐의 맨 앞 컴퓨터 확인
        q.pop(); // 큐에서 컴퓨터를 빼냄

        for(int neighbour : graph[current_computer])
        {
            if(!visited[neighbour]) // 만일 이웃 컴퓨터를 방문하지 않았을 경우
            {
                visited[neighbour] = true; // 방문 표기
                q.push(neighbour); // 큐에 넣음
                count_infected++; // 감염된 컴퓨터의 수를 1 증가시킴
            }
        }
    }
    
    cout << count_infected << "\n";

    return 0;
}

File: Manuel/2025-07-31/바이러스/BFS/virus_bfs.py
import sys
from collections import deque # 큐 구현을 위해 deque 사용

def solve():
    num_computers = int(sys.stdin.readline()) # 첫째 줄: 컴퓨터 수
    num_connections = int(sys.stdin.readline()) # 둘째 줄: 연결되어 있는 컴퓨터 쌍의 수
    graph = [[] for _ in range(num_computers + 1)] # 그래프를 인접 리스트로 표현; 컴퓨터의 번호를 1부터 N까지 매김

    for _ in range(num_connections): # 연결 정보를 그래프에 추가
        c1, c2 = map(int, sys.stdin.readline().split())
        # 양방향 연결
        graph[c1].append(c2)
        graph[c2].append(c1)
    
    # 방문 여부를 기록할 배열 (처음에는 모두 False로 초기화)
    visited = [False] * (num_computers + 1) # 컴퓨터의 번호가 1부터 시작하므로 배열은 (num_computers + 1) 크기로 만듦

    # BFS용 큐 생성
    q = deque()
    # 시작점인 1번 컴퓨터 넣기
    q.append(1)
    # 1번 컴퓨터는 방문 표시
    visited[1] = True

    infected_count = 0 # 1번을 제외한 바이러스에 걸리는 컴퓨터의 수

    while q: # BFS 탐색 시작
        current_computer = q.popleft() # 큐에서 현재 컴퓨터를 꺼냄

        # 현재 컴퓨터에 연결된 모든 이웃 컴퓨터 확인
        for neighbour in graph[current_computer]:
            if not visited[neighbour]: # 이웃 컴퓨터를 방문하지 않았다면
                visited[neighbour] = True # 방문했다고 표시
                q.append(neighbour) # 큐에 삽입
                infected_count += 1 # 감염된 컴퓨터 수 1 증가
        
    print(infected_count) # 감염된 컴퓨터 수 출력

solve()